test/e2e/app_secret_configmap_reconcile_test.go

```diff
@@ -0,0 +1,256 @@
+// Copyright 2021 VMware, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+package e2e
+
+import (
+	"fmt"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/ghodss/yaml"
+	"github.com/vmware-tanzu/carvel-kapp-controller/pkg/apis/kappctrl/v1alpha1"
+	corev1 "k8s.io/api/core/v1"
+)
+
+func Test_AppReconcileOccurs_WhenSecretUpdated(t *testing.T) {
+	env := BuildEnv(t)
+	logger := Logger{}
+	kapp := Kapp{t, env.Namespace, logger}
+	kubectl := Kubectl{t, env.Namespace, logger}
+	sas := ServiceAccounts{env.Namespace}
+
+	name := "configmap-with-secret"
+	// syncPeriod set to 1 hour so that test
+	// won't pass because of reconcile from time sync.
+	appYaml := fmt.Sprintf(`
+---
+apiVersion: kappctrl.k14s.io/v1alpha1
+kind: App
+metadata:
+  name: %s
+  annotations:
+    kapp.k14s.io/change-group: kappctrl-e2e.k14s.io/apps
+spec:
+  syncPeriod: 1h
+  serviceAccountName: kappctrl-e2e-ns-sa
+  fetch:
+    - inline:
+        paths:
+          file.yml: |
+            #@ load("@ytt:data", "data")
+            apiVersion: v1
+            kind: ConfigMap
+            metadata:
+              name: configmap
+            data:
+              hello_msg: #@ data.values.hello_msg
+  template:
+  - ytt:
+      inline:
+        pathsFrom:
+          - secretRef:
+              name: simple-app-values
+      paths:
+        - file.yml
+  deploy:
+    - kapp: {}
+---
+apiVersion: v1
+kind: Secret
+metadata:
+  name: simple-app-values
+stringData:
+  values.yml: |
+    #@data/values
+    ---
+    hello_msg: original`, name) + sas.ForNamespaceYAML()
+
+	cleanUp := func() {
+		kapp.Run([]string{"delete", "-a", name})
+	}
+	cleanUp()
+	defer cleanUp()
+
+	logger.Section("deploy", func() {
+		kapp.RunWithOpts([]string{"deploy", "-f", "-", "-a", name}, RunOpts{StdinReader: strings.NewReader(appYaml)})
+		out := kapp.Run([]string{"inspect", "-a", name, "--raw", "--tty=false", "--filter-kind=App"})
+
+		var cr v1alpha1.App
+		err := yaml.Unmarshal([]byte(out), &cr)
+		if err != nil {
+			t.Fatalf("Failed to unmarshal: %s", err)
+		}
+
+		// check for successful deployment
+		if cr.Status.Deploy == nil || !cr.Status.Deploy.Finished || cr.Status.Deploy.ExitCode != 0 {
+			t.Fatalf("Expected %s deployment to succeed but got:\n%s", name, cr.Status.Deploy.Stdout)
+		}
```

Comment:
kapp deploy technically waits for App CR to sucessfully reach ReconcileSucceeded state or otherwise fails if ReconcileFailed was reached.